src/ng/compile.js

```diff
@@ -1997,9 +1997,11 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
           for (i in elementControllers) {
             controller = elementControllers[i];
             var controllerResult = controller();
+
             if (controllerResult !== controller.instance) {
+              // Assign controller return value to the instance
               controller.instance = controllerResult;
-              $element.data('$' + directive.name + 'Controller', controllerResult);
+              $element.data('$' + i + 'Controller', controllerResult);
```

Comment:
Looks good to me, maybe the comment could be a bit more clear --- we're overwriting stuff that was done in `setupControllers()` because the return value was different from the pre-allocated instance.